lib-rpc/ocamlformat_rpc_lib.mli

```diff
@@ -9,70 +9,115 @@
 (*                                                                        *)
 (**************************************************************************)
 
-module type Command_S = sig
-  type t
+(** The [IO] module defines the blocking interface for reading and writing to
+    Cohttp streams *)
+module type IO = sig
+  (** ['a t] represents a blocking monad state *)
+  type +'a t
+
+  val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
+  (** [a >>= b] will pass the result of [a] to the [b] function. This is a
+      monadic [bind]. *)
+
+  val return : 'a -> 'a t
+  (** [return a] will construct a constant IO value. *)
+
+  (** [ic] represents an input channel *)
+  type ic
+
+  (** [oc] represents an output channel *)
+  type oc
+
+  val read_line : ic -> string option t
+  (** [read_line ic] will read a single line terminated by CR or CRLF from
+      the input channel [ic]. It returns {!None} if EOF or other error
+      condition is reached. *)
+
+  val read : ic -> int -> string t
+  (** [read ic len] will block until a maximum of [len] characters are read
+      from the input channel [ic]. It returns an empty string if EOF or some
+      other error condition occurs on the input channel, and can also return
+      fewer than [len] characters if input buffering is not sufficient to
+      satisfy the request. *)
+
+  val write : oc -> string -> unit t
```

Comment:
An alternative design is to make the primitives:

```
val read : ic -> Csexp.t option t
val write : oc -> Csexp.t list -> unit t (* the list allows us to control flushing a little better *)
val close : oc -> unit
```

The advantage of this scheme is that your code becomes simpler and users will be able to provide a more performant implementation than something that is byte based. The disadvantage of course is that you're leaking the serialization protocol.

Reply:
I changed the module signature and fixed the tests, it looks better indeed.